cpp/include/cudf/detail/utilities/device_atomics.cuh

```diff
@@ -116,18 +116,17 @@ struct genericAtomicOperationImpl<T, Op, 4> {
     using T_int = unsigned int;
 
     T old_value = *addr;
-    T assumed{old_value};
 
     do {
-      assumed           = old_value;
-      const T new_value = op(old_value, update_value);
+      const T_int assumed   = type_reinterpret<T_int, T>(old_value);
+      const T_int new_value = type_reinterpret<T_int, T>(op(old_value, update_value));
 
-      T_int ret = atomicCAS(reinterpret_cast<T_int*>(addr),
-                            type_reinterpret<T_int, T>(assumed),
-                            type_reinterpret<T_int, T>(new_value));
-      old_value = type_reinterpret<T, T_int>(ret);
+      const T_int ret = atomicCAS(reinterpret_cast<T_int*>(addr), assumed, new_value);
+      old_value       = type_reinterpret<T, T_int>(ret);
 
-    } while (assumed != old_value);
+      if (assumed == ret) { break; }
+
+    } while (true);
```

Comment:
I preferred the previous construction where `assumed` was declared first, and we used the `while` condition to break instead of an `if (condition) { break; }`. That's not part of the bug fix, so I'd prefer to revert that modification.

Reply:
Sure. Then `assume` will have to be declared out of block.

Reply:
I'll approve this PR now, but would like to see this reverted if it makes sense.

Reply:
Wait. That's more difficult: In addition to `assumed` to be declared out of block, another variable needs to be added to store `ret` out of block.